crates/saya/core/src/lib.rs

```diff
@@ -165,42 +166,144 @@ impl Saya {
                 continue;
             }
 
             let (last_state_root, params) =
-                self.prefetch_blocks(block..=latest_block, previous_block_state_root).await?;
+                self.prefetch_blocks(block..=maximum_expected, previous_block_state_root).await?;
             previous_block_state_root = last_state_root;
 
             // Updating the local state sequentially, as there is only one instance of
             // `self.blockchain` This part does no actual  proving, so should not be a
             // problem
-            for p in params {
-                self.process_block(&mut prove_scheduler, block, p)?;
 
-                if prove_scheduler.is_full() {
-                    self.process_proven(prove_scheduler).await?;
+            match self.config.mode {
+                SayaMode::PersistentMerging => {
+                    for p in params {
+                        let mut scheduler = if let Some(scheduler) = prove_scheduler {
+                            prove_scheduler = None;
+                            scheduler
+                        } else {
+                            Scheduler::new(
+                                self.config.batch_size,
+                                self.config.world_address,
+                                self.prover_identifier.clone(),
+                            )
+                        };
+
+                        if let Some((prover_input, _)) = self.process_block(block, p)? {
+                            scheduler.push_diff(prover_input)?;
+                            if scheduler.is_full() {
+                                self.process_proven(scheduler.proved().await?).await?;
+                            } else {
+                                prove_scheduler = Some(scheduler);
+                            }
+                        }
+
+                        block += 1;
+                    }
+                }
 
-                    prove_scheduler = Scheduler::new(
-                        self.config.batch_size,
-                        self.config.world_address,
-                        prover_identifier.clone(),
-                    );
+                SayaMode::Persistent => {
+                    let num_blocks = params.len() as u64;
+                    let calls = params
+                        .into_iter()
+                        .enumerate()
+                        .map(|(i, p)| self.process_block(block + i as u64, p))
+                        .collect::<Result<Vec<_>, _>>()?
+                        .into_iter()
+                        .flatten()
+                        .flat_map(|(_, c)| c)
+                        .map(|c| BatcherCall {
+                            to: c.to,
+                            selector: c.selector,
+                            calldata: c.calldata,
+                            starknet_messages: Vec::new(), // TODO: Fill messages.
```

Comment:
This should be done in a subsequent PR, or could be done in this one? The messages are or `L1HandlerTransaction` or available into the receipts.

Reply:
Should be trivial to implement, but might be less so to test it, that why I left it for now to avoid a false positive. 
This PR is quite big already so I would leave it for the following out, unless you think otherwise.

Reply:
Sounds great in a follow up PR. 👍